### ISSUE TITLE
Strange bug in read/show and type variables

### ISSUE THREAD

**Issue by johannesgerer** [ghcjs/ghcjs#762]
I am defining the two identical functions. However, they behave differently if I also define and use a more concrete version of the first function:

    module Main (main) where
    
    f1 :: (RealFloat a, Show a, Read a) => a -> a -> Bool
    f1 i j = read (show $ i/j) == i/j
    
    f2 :: (RealFloat a, Show a, Read a) =>  a -> a -> Bool
    f2 i j = read (show $ i/j) == i/j
    
    f1_breaker  :: Float -> Float -> Bool
    f1_breaker = f1
    
    main :: IO ()
    main = do
      putStr "expected: all three results identical"
      putStr "actual: "
      print $ [f1         0.1 (0.3::Float)
             , f1_breaker 0.1 0.3
             , f2         0.1 (0.3::Float)]
    
Compiling the above with GHCJS version 8.6.0.1 (GHC 8.6.5) and running it in node v10.16.3 and Chrome 77 gives:

    > ghcjs /tmp/f.hs -O1 && node /tmp/f.jsexe/all.js

    expected: all three results identical
    actual: [False,False,True]

PS: without optimization it works fine

    > ghcjs /tmp/f.hs && node /tmp/f.jsexe/all.js

    expected: all three results identical
    actual: [False,False,False]

**Comment by cdsmith** [comment]
I'm not sure I understand.  It looks like `f1`, `f2`, and `f1_breaker` are all consistent in evaluating to `False`

As for why `read . show` is not the identity, I suspect it has to do with the fact that GHCJS always uses `Double` under the hood, since JavaScript doesn't have a native single-precision floating point type.  So `read` and  `show` instances for `Float` probably throw away the extra precision, and the result isn't equal.

**Comment by johannesgerer** [comment]
Yes, sorry for the confusion. I pasted the wrong output. Please see the updated question with clarified compiler flags.

**Comment by johannesgerer** [comment]
So to clarify: It looks like a bug.

`f2` returns `True` whereas the other two functions which are identical to `f2` return `False`.
